internal/signerverifier/ssh/ssh.go

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+package ssh
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"encoding/base64"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/hiddeco/sshsig"
+	"golang.org/x/crypto/ssh"
+)
+
+// SSH Key/Verifier
+type Key struct {
+	KeyType string
+	KeyVal  KeyVal
+	Scheme  string
+	keyID   string
+}
+
+type KeyVal struct {
+	Public string
+}
+
+// SSH Key.Verify() implementation of dsse.Verifier interface.
+func (k *Key) Verify(_ context.Context, data []byte, sig []byte) error {
+	pub, err := parseSSH2Body(k.KeyVal.Public)
+	if err != nil {
+		return fmt.Errorf("failed to parse ssh public key material: %w", err)
+	}
+
+	signature, err := sshsig.Unarmor(sig)
+	if err != nil {
+		return fmt.Errorf("failed to parse ssh signature: %w", err)
+	}
+
+	message := bytes.NewReader(data)
+
+	// ssh-keygen uses sha512 to sign with **any*** key
+	hash := sshsig.HashSHA512
+	if err = sshsig.Verify(message, signature, pub, hash, "gittuf"); err != nil {
+		return fmt.Errorf("failed to verify ssh signature: %w", err)
+	}
+
+	return nil
+}
+
+// SSH Key.KeyID() implementation of dsse.Verifier interface.
+// FIXME: consider removing error in interface; the dsse implementation
+// clearly needs one
+func (k *Key) KeyID() (string, error) {
+	return k.keyID, nil
+}
+
+// SSH Key.Public() implementation of dsse.Verifier interface.
+// FIXME: consider removing in interface, "Verify()" is all that's needed
+func (k *Key) Public() crypto.PublicKey {
+	sshKey, _ := parseSSH2Body(k.KeyVal.Public)
+	return sshKey.(ssh.CryptoPublicKey).CryptoPublicKey()
+}
+
+// SSH Signer
+type Signer struct {
+	Key  *Key
+	Path string
+}
+
+// SSH Signer.Sign() implementation of dsse.Signer interface.
+// Signs using "s.Path" to a public or private, encrypted or plaintext, rsa,
+// ecdsa or ed25519 key file supported by "ssh-keygen". This aligns with the
+// git "user.signingKey" option.
+// https://git-scm.com/docs/git-config#Documentation/git-config.txt-usersigningKey
+func (s *Signer) Sign(_ context.Context, data []byte) ([]byte, error) {
+	cmd := exec.Command("ssh-keygen", "-Y", "sign", "-n", "gittuf", "-f", s.Path) //nolint:gosec
+
+	cmd.Stdin = bytes.NewBuffer(data)
+
+	output, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to run command %v: %w", cmd, err)
+	}
+
+	return output, nil
```

Comment:
So this implements the DSSE.Signer interface, meaning the signature returned here is meant to be embedded in a DSSE envelope primarily. Does this mean we should format the signature here to strip the header and footer? Remove newlines etc? DSSE signature should be b64 encoded which the rest of this is.

Or, we leave it as is and effectively double b64 encode it: https://github.com/gittuf/gittuf/blob/main/internal/signerverifier/dsse/dsse.go#L55.

Reply:
I lean towards the latter. The DSSE Signer.sign interface asks me to return bytes, so I return bytes. 

Because even if I returned the b64 encoded bytes only, with begin/end marker and newlines stripped, then the current DSSE implementation would still re-b64-encode them.

So, I'd have to strip all that stuff and b64-decode the bytes before I return them, which sounds like a lot of work and room for error. 

Alternatively, we can also change the dsse Signer interface upstream.